util/reggen/gen_tock.py

```diff
@@ -0,0 +1,384 @@
+        fieldout: TextIO,
+        reg: Any,  # Should be Register|IpBlock
+        name: str,
+        width: int) -> None:
+    genout(fieldout, "{} {} [", FIELD_VISIBILITY, name.upper())
+    fields: Any
+    if isinstance(reg, Register):
+        fields = reg.fields
+    elif isinstance(reg, IpBlock):
+        fields = reg.interrupts
+    else:
+        raise TypeError(type(reg))
+
+    for field in fields:
+        if field.bits.width() == width:
+            genout(fieldout, "],\n")
+            break
+        genout(fieldout, "\n{} OFFSET({}) NUMBITS({}) [",
+               field.name.upper(), field.bits.lsb, field.bits.width())
+        if getattr(field, 'enum', None) is not None:
+            for enum in field.enum:
+                genout(fieldout, "\n{} = {},", enum.name.upper(), enum.value)
+            genout(fieldout, "\n],")
+        else:
+            genout(fieldout, "],")
+    else:
+        genout(fieldout, "\n],\n")
+
+
+def gen_const_register(regout: TextIO, fieldout: TextIO, reg: Register,
+                       block: IpBlock, rnames: Set[str],
+                       existing_defines: Set[str], access_type: Set[str],
+                       highest_address: Set[int]) -> None:
+    if not block.no_auto_intr and reg.name.startswith('INTR_'):
+        rname = "INTR"
+    else:
+        rname = reg.name.upper()
+        gen_field_definitions(fieldout, reg, rname, block.regwidth)
+
+    a = access(reg)
+    access_type.add(a)
+    highest_address.add(reg.offset + block.regwidth // 8)
+    genout(regout, format_comment(first_line(reg.desc)))
+    genout(regout, "(0x{:04x} => {} {}: {}<u32, {}::Register>),\n", reg.offset,
+           REG_VISIBILITY, reg.name.lower(), a, rname)
+
+
+def gen_const_window(regout: TextIO, win: Window, block: IpBlock,
+                     rnames: Set[str], existing_defines: Set[str],
+                     access_type: Set[str], highest_address: Set[int]) -> None:
+    genout(regout, format_comment('Memory area: ' + first_line(win.desc)))
+    a = access(win)
+    access_type.add(a)
+    highest_address.add(win.offset + win.items * block.regwidth // 8)
+    genout(regout, "(0x{:04x} => {} {}: [{}<u32>; {}]),\n", win.offset,
+           REG_VISIBILITY, win.name.lower(), a, win.items)
+
+
+def gen_rust_module_param(outstr: TextIO, param: LocalParam, module_name: str,
+                          existing_defines: Set[str]) -> None:
+    # Presently there is only one type (int), however if the new types are
+    # added, they potentially need to be handled differently.
+    known_types = ["int"]
+    if param.param_type not in known_types:
+        warnings.warn("Cannot generate a module define of type {}".format(
+            param.param_type))
+        return
+
+    if param.desc is not None:
+        genout(outstr, format_comment(first_line(param.desc)))
+    # Heuristic: if the name already has underscores, it's already snake_case,
+    # otherwise, assume StudlyCaps and covert it to snake_case.
```

Comment:
nit: "convert" is misspelled as "covert" here.

Reply:
Done.